### ISSUE TITLE
Crash when blocks are deallocated and have strong references to views

### ISSUE THREAD

**Issue by steipete** [pokeb/asi-http-request#91]
I'm having crashes when blocks retain elements and are released in own threads.

 bool _WebTryThreadLock(bool), 0x7d7c3e0: Tried to obtain the web lock from a thread other than the main thread or the web thread. This may be a result of calling to UIKit from a secondary thread. Crashing now... (EXC_BAD_ACCESS)

http://skitch.com/steipetee/d9xwx/system

**Comment by pokeb** [comment]
Hi Peter

Hmm. Perhaps the block release code needs to move over to the main thread. If I can reproduce I'll try this today.

Best,

Ben

**Comment by steipete** [comment]
reproducing is not that easy, best you create a view, remove if from superview (so it should get autoreleaed) but keep a strong reference in the block, so it's going to be removed from the block in a worker thread.

**Comment by pokeb** [comment]
Release blocks on main thread (hopefully closed by 341fb517c0793ba66191a1c2974faf289aba50ad)

**Comment by pokeb** [comment]
Hi again

Can you let me know if this solves the issue for you?

Best,

Ben

**Comment by steipete** [comment]
Thanks! I'll reopen it if this ever occurs again.

**Comment by alanzeino** [comment]
I got the same issue with the latest version of ASIHTTPRequest since I use ASINetworkQueue. Luckily in Xcode 4 I was able to see that after posting an NSNotification back to a View Controller, a call to UIKit was being made on the same thread upon which requestFinished was called (I override requestFinished and failWithError:, and post the NSNotification there). The fix is to use GCD to ensure the call to UIKit is made on the main queue in the View Controller:

dispatch_async(dispatch_get_main_queue(), ^{
    [self dismissModalViewControllerAnimated:YES];
 });

So in other words, if you override the success/fail methods in a subclass of ASIHTTPRequest (which exists in an ASINetworkQueue, et al) and then fire an NSNotification back to a View Controller, you may find that subsequent methods will fire in a queue not in the main thread.

**Comment by pokeb** [comment]
Hi Alan

I guess you could override reportFinished/reportFailure instead - they should always be called on the main thread.

Best,

Ben
